### PR TITLE
Add Mining Levels and Tool Attributes API

### PR DESCRIPTION
**Pull request by NoComment1105**
This pull request ports the Mining Level API and the Tool Attributes API from the Fabric API to Quilt Standard Libararies.
In its current state it is a near direct port, with only minor changes to some class names, due to other QSL naming.
I am opening this pull-request as a draft for now, because i wish to rewrite certain parts of this, I am aware other wish to do this too, but I don't want to do anything bad, so it's here for people to let me know when I've done somethings wrong

### CODE REVIEW

File: library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/BooleanFunction.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package org.quiltmc.qsl.base.api.util;
+
+/**
+ * Represents a function that accepts an boolean-valued argument and produces a result.
+ *
+ * <p>This is the {@code boolean}-consuming primitive specialization for {@link java.util.function.Function}.
+ */
+@FunctionalInterface
+public interface BooleanFunction<R> {
```
Comment (OroArmor) [inline library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/BooleanFunction.java]: Idea: extend `Function<Boolean, R>`

File: library/item/tool_attributes/src/main/java/org/quiltmc/qsl/tool_attributes/api/QuiltToolTags.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package org.quiltmc.qsl.tool_attributes.api;
+
+import net.minecraft.item.Item;
+import net.minecraft.tag.Tag;
+import net.minecraft.util.Identifier;
+import org.quiltmc.qsl.tag.api.TagRegistry;
+
+public class QuiltToolTags {
```
Comment (OroArmor) [inline library/item/tool_attributes/src/main/java/org/quiltmc/qsl/tool_attributes/api/QuiltToolTags.java:28]: Javadocs probably wanted

**Review by OroArmor — APPROVED** [review]
Looks very good so far.

File: library/item/tool_attributes/src/main/java/org/quiltmc/qsl/tool_attributes/api/QuiltToolTags.java (hunk 2 above)
Comment (NoComment1105) [reply library/item/tool_attributes/src/main/java/org/quiltmc/qsl/tool_attributes/api/QuiltToolTags.java:28]: Added

File: library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/BooleanFunction.java (hunk 1 above)
Comment (NoComment1105) [reply library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/BooleanFunction.java]: done

**Comment by EnnuiL** [comment]
It has been, like, more than a year since this PR was opened; I'll be closing this

**Pull request closed without merging**